### PR TITLE
PyTorch/Caffe2 Build Path Changes (ROCm path) + MIOpen Integration

### PR DESCRIPTION
**Pull request by Jorghi12**
This Pull Request includes the following:
a) Integration with [AMD's MIOpen library ](https://github.com/ROCmSoftwarePlatform/MIOpen) (think CuDNN).
b) Sharing more of the ROCm build path between Caffe2/PyTorch.
c) Creating libcaffe2_hip.so

### CODE REVIEW

File: aten/src/ATen/CMakeLists.txt [hunk 1]
```diff
@@ -371,6 +389,9 @@ if(NOT AT_LINK_STYLE STREQUAL "INTERFACE")
     if(NOT NO_CUDA OR USE_ROCM)
       set_property(TARGET ATen_cuda PROPERTY CXX_STANDARD 11)
     endif()
+    if(USE_ROCM)
+      set_target_properties(ATen_cuda PROPERTIES LINKER_LANGUAGE HIP)
```
Comment (ezyang) [inline aten/src/ATen/CMakeLists.txt]: Don't forget that eventually we want `ATen_rocm/hip` to be a thing :)

File: aten/src/ATen/miopen/Descriptors.cpp [hunk 2]
```diff
@@ -0,0 +1,120 @@
+#include "Descriptors.h"
+#include <ATen/ATen.h>
+
+namespace at { namespace native {
```
Comment (ezyang) [inline aten/src/ATen/miopen/Descriptors.cpp]: So, I remember a week ago, you told me that `miopen` was API compatible with `cudnn`, and so we were going to regex our way to victory. In this PR, I see a copy of code for `miopen`, which is basically copied straight from the existing cudnn bindings. What constraints changed that lead us here?

Comment (ezyang) [reply aten/src/ATen/miopen/Descriptors.cpp]: If we're going to actually do the "fork the code and maintain it as actual source", we had better also fix up all the references to CuDNN in the comments...

File: aten/src/ATen/miopen/Descriptors.h [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+descriptor once and cache it.  However, our ATen interface is (1) stateless (so
+we can't cache the descriptors) and (2) does not accept arbitrary user types in
+its interface (so we can't pass the descriptor in).  This puts us in a pickle.
+
+In cuDNN 7, a new function, cudnnRestoreDropoutDescriptor was added, which
+forgoes the expensive initialization process, and can initialize the
+descriptor with a pre-initialized state CUDA tensor.  This is great, because
+it means we can simply pass in the state tensor and then initialize the
+descriptor internally.  Unfortunately, this function is not available in
+cuDNN 6.
```
Comment (ezyang) [inline aten/src/ATen/miopen/Descriptors.h]: No dropout support, so this comment is irrelevant...

File: aten/src/ATen/miopen/Descriptors.h [hunk 4]
```diff
@@ -0,0 +1,323 @@
...
+
+struct ConvolutionDescriptor
+  : public Descriptor<miopenConvolutionDescriptor,
+                      &miopenCreateConvolutionDescriptor,
+                      &miopenDestroyConvolutionDescriptor>
+{
+  void set(miopenDataType_t dataType, int dim, int* pad, int* stride, int * upscale /* aka dilation */, int groups) {
+    miopenDataType_t mathType = dataType;
+    if (dataType == miopenHalf) mathType = miopenFloat;
+    //?????????????? What the fuck? It's asking for the stride for the height & stride for the width seperately....!
```
Comment (ezyang) [inline aten/src/ATen/miopen/Descriptors.h]: Is it supporting asymmetric kernels?

File: aten/src/ATen/miopen/Descriptors.h [hunk 5]
```diff
@@ -0,0 +1,323 @@
...
+  }
+};
+
+struct SpatialTransformerDescriptor
+  : public Descriptor<miopenSpatialTransformerStruct,
+                      &miopenCreateSpatialTransformerDescriptor,
+                      &miopenDestroySpatialTransformerDescriptor>
+{
+  void set(miopenDataType_t dataType, int dim, int* size) {
+    //?????????????? This function doesn't even exist!
```
Comment (ezyang) [inline aten/src/ATen/miopen/Descriptors.h]: What function?

File: aten/src/ATen/miopen/Utils.h [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+#include <ATen/ATen.h>
+#include "THC/THC.h"
+#include "miopen-wrapper.h"
+#include "Handles.h"
+
+namespace at { namespace native {
+
+inline void setMIOpenStreamToCurrent() {
+  // TODO: Should getCurrentStream be a method on Context?
+  MIOPEN_CHECK(miopenSetStream(getMiopenHandle(), THCState_getCurrentStream(globalContext().getTHCState())));
```
Comment (ezyang) [inline aten/src/ATen/miopen/Utils.h]: It's a method on context now, so please fix this call site :)

File: aten/src/ATen/miopen/Utils.h [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+}
+
+// cuDNN has a buggy check for tensor being contiguous (that is, it does
+// not ignore stride for dimension that is equal to 0).  This function
+// makes tensors which have zero stride contiguous, by setting the
+// strides to 1 as cuDNN likes.
```
Comment (ezyang) [inline aten/src/ATen/miopen/Utils.h]: Is miopen bug-for-bug compatible here?

File: aten/src/ATen/native/cudnn/Conv.cpp [hunk 8]
```diff
@@ -323,7 +323,7 @@ struct ConvolutionArgs {
 
 // Hashing machinery for ConvolutionParams
 struct ParamsHash {
-  size_t operator()(const ConvolutionParams& params) const {
+  std::size_t operator()(const ConvolutionParams& params) const {
```
Comment (ezyang) [inline aten/src/ATen/native/cudnn/Conv.cpp]: @goldsborough got rid of the `std::` qualifiers. Does the HIP compiler need the qualifier here? Does including `<cstddef>` help here?

File: aten/src/ATen/native/miopen/BatchNorm.cpp [hunk 9]
```diff
@@ -0,0 +1,213 @@
...
+#include <ATen/NativeFunctions.h>
+#include <ATen/Config.h>
+
+#if !AT_MIOPEN_ENABLED
+
+namespace at { namespace native {
+
+// See Note [ATen preprocessor philosophy]
+
+std::tuple<Tensor, Tensor, Tensor> cudnn_batch_norm(
```
Comment (ezyang) [inline aten/src/ATen/native/miopen/BatchNorm.cpp]: cudnn?

File: aten/src/ATen/native/miopen/Conv.cpp [hunk 10]
```diff
@@ -0,0 +1,1209 @@
...
+        cudnn_convolution_BWD_DATA_PREFER_FASTEST,
+        0,
+        algo));*/
+  }
+
+  static void getWorkspaceSize(
+    const ConvolutionArgs& args,
+    miopenConvBwdDataAlgorithm_t algo, size_t* workspaceSize)
+  {
+    throw std::runtime_error("miopenGetConvolutionBackwardDataWorkspaceSize is not supported.");
```
Comment (ezyang) [inline aten/src/ATen/native/miopen/Conv.cpp]: These workspace functions are all internal ones, so if miopen doesn't need them, get rid of them.

File: setup.py [hunk 11]
```diff
@@ -639,14 +651,18 @@ def run(self):
...
+        CAFFE2_LIBS.append(os.path.join(lib_path, 'libcaffe2_hip.dylib'))
     NCCL_LIB = os.path.join(lib_path, 'libnccl.1.dylib')
 
 if IS_WINDOWS:
     CAFFE2_LIBS = [os.path.join(lib_path, 'caffe2.lib')]
-    if WITH_CUDA or WITH_ROCM:
+    if WITH_CUDA:
         CAFFE2_LIBS.append(os.path.join(lib_path, 'caffe2_gpu.lib'))
+    if WITH_ROCM:
+        CAFFE2_LIBS.append(os.path.join(lib_path, 'caffe2_hip.lib'))
```
Comment (ezyang) [inline setup.py]: ...oh? How does this work?

**Comment by ezyang** [comment]
@Jorghi12 I'd strongly advise splitting this PR into the build system bit, and the miopen bit. We will be able to land the build system a bit quicker; I want to figure out if we should be regexing miopen or not; if we're not, I have a lot more comments...

**Comment by dagamayank** [comment]
Is this PR still valid?

**Comment by yf225** [comment]
@Jorghi12 What should be the next step for this PR?

**Comment by iotamudelta** [comment]
Yes, I agree that it has been overcome by events. It'd be great if @Jorghi12 could check if there is and if so salvage whatever is not in #8812 or [ROCm-PyTorch#22](https://github.com/ROCmSoftwarePlatform/pytorch/pull/22) and create a PR against ROCm-PyTorch for staging. Thanks!

**Comment by bddppq** [comment]
Closing stale PR
